src/workers/parser/mapBindings.js

```diff
@@ -10,18 +10,95 @@ import { isTopLevel } from "./utils/helpers";
 import generate from "@babel/generator";
 import * as t from "@babel/types";
 
+function getIdentifierName(identifier, bindings) {
+  return bindings.includes(identifier.name)
+    ? identifier
+    : t.memberExpression(t.identifier("self"), identifier);
+}
+
+function destructurize(body, bindings, baseObj, property) {
+  let name = property.value;
+  let value = null;
+
+  if (t.isAssignmentPattern(property.value)) {
+    name = property.value.left;
+
+    value = t.conditionalExpression(
+      t.binaryExpression(
+        "===",
+        t.memberExpression(baseObj, property.key),
+        t.identifier("undefined")
+      ),
+      property.value.right,
+      t.memberExpression(baseObj, property.key)
+    );
+  }
+
+  if (t.isIdentifier(name)) {
+    body.push(
+      t.assignmentExpression(
+        "=",
+        getIdentifierName(name, bindings),
+        value === null ? t.memberExpression(baseObj, property.key) : value
+      )
+    );
+  } else if (t.isObjectPattern(name)) {
+    const identifier =
+      value === null
+        ? baseObj
+        : t.identifier(
+            `${baseObj.object ? baseObj.object.name : baseObj.name}${
+              property.key.name
+            }__`
+          );
+
+    if (value !== null) {
+      body.push(
+        t.variableDeclaration("let", [t.variableDeclarator(identifier, value)])
+      );
+    }
+
+    for (const childProperty of name.properties) {
+      destructurize(
+        body,
+        bindings,
+        value === null
+          ? t.memberExpression(identifier, property.key)
+          : identifier,
+        childProperty
+      );
+    }
+  }
+}
+
 // translates new bindings `var a = 3` into `self.a = 3`
 // and existing bindings `var a = 3` into `a = 3` for re-assignments
 function globalizeDeclaration(node, bindings) {
-  return node.declarations.map(declaration => {
-    const identifier = bindings.includes(declaration.id.name)
-      ? declaration.id
-      : t.memberExpression(t.identifier("self"), declaration.id);
+  return node.declarations.reduce((body, declaration, i) => {
+    if (t.isPattern(declaration.id)) {
```

Comment:
Clearly a lot of work went into this, but unless I'm missing something, I don't think we strictly need to do so much to implement this. For instance, for
```
expression: "const { a, c: y } = await b()",
newExpression: formatAwait(`
  let __decl0__ = await b();
  self.a = __decl0__.a;
  return (self.y = __decl0__.c);
`)
```
we could do
```
expression: "const { a, c: y } = await b()",
newExpression: formatAwait(`
  ({ a: self.a, c: self.y } = await b());
`)
```
to avoid re-implementing object destructuring, and also allow us to handle array pattern destructuring.

Reply:
Well, I just adapted the solution to the proposed output. Frankly speaking, I had a similar feeling during implementation. Nonetheless, I changed the implementation to preserve destructuring, as the code is way cleaner now and we can easily support array destructuring.